File: Algorithms/322. Coin Change.rs
use std::cmp;

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let length = (amount + 1) as usize;
        let mut dp = vec![-1; length];
        dp[0] = 0;
        
        for i in 1..length {
            for &coin in &coins {
                if coin as usize <= i {
                    let j = i - coin as usize;
                    if dp[j] != -1 {
                        if dp[i] == -1 {
                            dp[i] = dp[j] + 1
                        } else {
                            dp[i] = cmp::min(dp[i], dp[j] + 1);
                        }
                    }
                }
            }
        }
        dp[amount as usize]
    }
}

File: Algorithms/322. Coin Change.ts
function coinChange(coins: number[], amount: number): number {
  if (amount === 0) {
    return 0;
  }
  const length: number = amount + 1;
  const dp: number[] = new Array(length).fill(-1);
  dp[0] = 0;

  for (let i = 1; i < length; i++) {
    for (let coin of coins) {
      if (coin <= i) {
        let j: number = i - coin;
        if (dp[j] !== -1) {
          if (dp[i] === -1) {
            dp[i] = dp[j] + 1;
          } else {
            dp[i] = Math.min(dp[i], dp[j] + 1);
          }
        }
      }
    }
  }
  return dp[amount];
}
